lms/static/scripts/frontend_apps/components/ComponentWithLoaderWrapper.tsx

```diff
@@ -0,0 +1,60 @@
+import type { ComponentChildren } from 'preact';
+import { useEffect, useState } from 'preact/hooks';
+import { useParams, useRoute } from 'wouter-preact';
+
+import type { ConfigObject, Ensure } from '../config';
+import { useConfig } from '../config';
+
+export type LoaderOptions = {
+  config: Ensure<ConfigObject, 'dashboard' | 'api'>;
+  params: Record<string, string>;
+  signal: AbortSignal;
+};
+
+export default function ComponentWithLoaderWrapper() {
+  const config = useConfig(['dashboard', 'api']);
+  const [component, setComponent] = useState<ComponentChildren>();
+  const [loading, setLoading] = useState(true);
+  const [isAssignment, assignmentParams] = useRoute(
+    '/assignments/:assignmentId',
+  );
+  const [isCourse, courseParams] = useRoute('/courses/:courseId');
+  const [isHome] = useRoute('');
+  const globalParams = useParams();
+  const assignmentId = assignmentParams?.assignmentId ?? '';
+  const courseId = courseParams?.courseId ?? '';
+  const organizationId = globalParams.organizationId ?? '';
+
+  useEffect(() => {
+    const loaderModule = isAssignment
+      ? import('./dashboard/AssignmentActivity')
+      : isCourse
+        ? import('./dashboard/CourseActivity')
+        : import('./dashboard/OrganizationActivity');
+    const params = { assignmentId, courseId, organizationId };
+
+    const abortController = new AbortController();
+    loaderModule.then(async ({ loader, default: Component }) => {
+      // TODO Error handling
+      setLoading(true);
+      const loadResult = await loader({
+        config,
+        params,
+        signal: abortController.signal,
+      });
+      setLoading(false);
+      setComponent(<Component loadResult={loadResult} params={params} />);
+    });
+
+    return () => abortController.abort();
+  }, [assignmentId, courseId, organizationId, config, isAssignment, isCourse]);
+
+  return !component ? (
+    <>Initial loading...</>
+  ) : (
+    <>
+      {loading && <>Transitioning...</>}
+      {component}
+    </>
```

Comment:
This is just temporary, for demonstration purposes.

It represents the fact that we can have an initial loading state, a component that renders as soon as we have it, and an additional loader that can be shown when transitioning between routes, next to the "old" route until the data for the new route has been loaded.